src/elbo.jl

```diff
@@ -0,0 +1,247 @@
+"Likelihoods which take a scalar as input and return a scalar."
+ScalarLikelihood = Union{
+    BernoulliLikelihood,
+    PoissonLikelihood,
+    GaussianLikelihood,
+    ExponentialLikelihood,
+    GammaLikelihood
+}
+
+
+abstract type ExpectationMethod end
+struct Default <: ExpectationMethod end
+struct Analytic <: ExpectationMethod end
+
+struct Quadrature <: ExpectationMethod
+    n_points::Int
+end
+Quadrature() = Quadrature(20)
+
+struct MonteCarlo <: ExpectationMethod
+    n_samples::Int
+end
+MonteCarlo() = MonteCarlo(20)
+
+"""
+    elbo(fx::FiniteGP, y::AbstractVector{<:Real}, fz::FiniteGP, q::AbstractMvNormal; n_data=length(y), method=Default())
+
+Compute the Evidence Lower BOund from [1] for the process `fx.f` where `y` are
+observations of `fx`, pseudo-inputs are given by `z = fz.x` and `q(u)` is a
+variational distribution over inducing points `u = f(z)`.
+
+`method` selects which method is used to calculate the expected loglikelihood in
+the ELBO. The options are: `Default()`, `Analytic()`, `Quadrature()` and
+`MonteCarlo()`. For likelihoods with an analytic solution, `Default()` uses this
+exact solution. If there is no such solution, `Default()` either uses
+`Quadrature()` or `MonteCarlo()`, depending on the likelihood.
+
+N.B. the observation noise `fx.Σy` is assumed to be homoscedastic and
+uncorrelated - i.e. only `fx.Σy[1]` is used.
+
+[1] - Hensman, James, Alexander Matthews, and Zoubin Ghahramani. "Scalable
+variational Gaussian process classification." Artificial Intelligence and
+Statistics. PMLR, 2015.
+"""
+function AbstractGPs.elbo(
+    fx::FiniteGP{<:AbstractGP, <:AbstractVector, <:Diagonal{<:Real, <:Fill}},
+    y::AbstractVector{<:Real},
+    fz::FiniteGP,
+    q::AbstractMvNormal;
+    n_data=length(y),
+    method=Default()
+)
+    return _elbo(method, fx, y, fz, q, GaussianLikelihood(fx.Σy[1]), n_data)
+end
+
+function AbstractGPs.elbo(::FiniteGP, ::AbstractVector, ::FiniteGP, ::AbstractMvNormal; kwargs...) where T<:FiniteGP
+    return error("The observation noise fx.Σy may not be homoscedastic.\n To avoid this error, construct fx using: f = GP(kernel); fx = f(x, σ²)")
```

Comment:
Do you mean heteroscedastic here?
```suggestion
    return error("The observation noise fx.Σy may not be heteroscedastic.\n To avoid this error, construct fx using: f = GP(kernel); fx = f(x, σ²)")
```

Reply:
I meant it as in "the noise you supplied might not be homoscedastic, you should fix it", but yeah the phrasing's terrible - what about `The observation noise fx.Σy must be homoscedastic...`?

Reply:
Ohhh okay. Yeah, that sounds good to me.